pkg/engine/context/deferred.go

```diff
@@ -0,0 +1,109 @@
+package context
+
+import (
+	"regexp"
+)
+
+type deferredLoader struct {
+	name    string
+	matcher regexp.Regexp
+	loader  Loader
+}
+
+func NewDeferredLoader(name string, loader Loader) (DeferredLoader, error) {
+	// match on ASCII word boundaries except do not allow starting with a `.`
+	// this allows `x` to match `x.y` but not `y.x` or `y.x.z`
+	matcher, err := regexp.Compile(`(?:\A|\z|\s|[^.0-9A-Za-z])` + name + `\b`)
+	if err != nil {
+		return nil, err
+	}
+
+	return &deferredLoader{
+		name:    name,
+		matcher: *matcher,
+		loader:  loader,
+	}, nil
+}
+
+func (dl *deferredLoader) Name() string {
+	return dl.name
+}
+
+func (dl *deferredLoader) HasLoaded() bool {
+	return dl.loader.HasLoaded()
+}
+
+func (dl *deferredLoader) LoadData() error {
+	return dl.loader.LoadData()
+}
+
+func (d *deferredLoader) Matches(query string) bool {
+	return d.matcher.MatchString(query)
+}
+
+type byLevel struct {
+	level  int
+	loader DeferredLoader
+}
+
+type deferredLoaders struct {
+	enableDeferredLoading bool
+	loaders               []byLevel
+}
+
+func NewDeferredLoaders(enableDeferredLoading bool) DeferredLoaders {
+	return &deferredLoaders{
+		enableDeferredLoading: enableDeferredLoading,
+		loaders:               make([]byLevel, 0),
+	}
+}
+
+func (d *deferredLoaders) Enabled() bool {
+	return d.enableDeferredLoading
+}
+
+func (d *deferredLoaders) Add(dl DeferredLoader, level int) {
+	d.loaders = append(d.loaders, byLevel{level, dl})
+}
+
+func (d *deferredLoaders) Reset(removeCheckpoint bool, level int) {
+	for i := len(d.loaders) - 1; i >= 0; i-- {
+		if d.loaders[i].level > level {
+			// remove loaders from a nested context (level > current)
+			d.loaders = append(d.loaders[:i], d.loaders[i+1:]...)
+		} else {
+			if d.loaders[i].loader.HasLoaded() {
+				// reload data into the current context
+				if err := d.loaders[i].loader.LoadData(); err != nil {
+					logger.Error(err, "failed to reload context entry", "name", d.loaders[i].loader.Name())
+				}
+				if d.loaders[i].level == level {
+					d.loaders = append(d.loaders[:i], d.loaders[i+1:]...)
+				}
+			}
+			if !removeCheckpoint {
```

Comment:
why `removeCheckpoint` should influence this ? i mean, removing or not shouldn't change the behaviour when restoring the context.

Reply:
I was looking for a better name than `removeCheckpoint`.  The value indicates whether the operation is a `restore` or a `reset`. Switched it to `restore`.

I thought more about a stack-like structure, and decided against it. Although it may simplify the reset behaviors, it will complicate the query. A list feels the right abstraction here, as new loaders are added to the end and a query needs to process all matching loaders and handle override behaviors.  

For example, if a nested context hides a var with the same name, its value should be returned while it is active. To do this we have to search across checkpoints i.e. treat the declarations as an ordered list.

I tightened up the logic and also added a guard for cases where a Reset/Restore is invoked without a checkpoint. It seems like this may be happening in the cleanup handler, as there were test case failures with the prior change.

Reply:
Not necessarily, every checkpoint can contain the whole list, no need to go through all checkpoints.

Reply:
I tried your latest changes and failed to find a case producing invalid results. I’ll try again tomorrow :)